src/oidc_child/oidc_child_curl.c

```diff
@@ -0,0 +1,462 @@
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to set http options.\n");
+        goto done;
+    }
+
+    res = curl_easy_perform(curl_ctx);
+    if (res != CURLE_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to send request.\n");
+        ret = EIO;
+        goto done;
+    }
+
+    res = curl_easy_getinfo(curl_ctx, CURLINFO_RESPONSE_CODE, &resp_code);
+    if (res != CURLE_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to get response code.\n");
+        ret = EIO;
+        goto done;
+    }
+
+    if (resp_code != 200) {
+        DEBUG(SSSDBG_OP_FAILURE, "Request failed, response code is [%ld].\n",
+                                 resp_code);
+        ret = EIO;
+        goto done;
+    }
+
+    ret = EOK;
+done:
+    curl_slist_free_all(headers);
+    curl_easy_cleanup(curl_ctx);
+    return ret;
+}
+
+#define AZURE_EXPECT_CODE "The request body must contain the following parameter: 'code'."
+
+errno_t get_token(TALLOC_CTX *mem_ctx,
+                  struct devicecode_ctx *dc_ctx, const char *client_id,
+                  const char *client_secret,
+                  bool get_device_code)
+{
+    CURL *curl_ctx = NULL;
+    CURLcode res;
+    int ret;
+    size_t waiting_time = 0;
+    char *error_description = NULL;
+    char *post_data = NULL;
+    const char *post_data_tmpl = "grant_type=urn:ietf:params:oauth:grant-type:device_code&client_id=%s&%s=%s";
+    struct curl_slist *headers = NULL;
+    bool azure_fallback = false;
+
+    headers = curl_slist_append(headers, "Accept: application/json");
+
+    post_data = talloc_asprintf(mem_ctx, post_data_tmpl, client_id, "device_code",
+                                                         dc_ctx->device_code);
+    if (post_data == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to generate POST data.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    if (client_secret != NULL) {
+        post_data = talloc_asprintf_append(post_data, "&client_secret=%s",
+                                           client_secret);
+        if (post_data == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "Failed to add client secret to POST data.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+    }
+
+    curl_ctx = curl_easy_init();
+    if (curl_ctx == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to initialize curl.\n");
+        ret = EIO;
+        goto done;
+    }
+
+    ret = set_http_opts(curl_ctx, dc_ctx, dc_ctx->token_endpoint, post_data,
+                        NULL, headers);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to set http options.\n");
+        goto done;
+    }
+
+    do {
+        clean_http_data(dc_ctx);
+
+        res = curl_easy_perform(curl_ctx);
+        if (res != CURLE_OK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to send token request.\n");
+            ret = EIO;
+            goto done;
+        }
+
+        ret = parse_token_result(dc_ctx, &error_description);
```

Comment:
This can leak `error_description` allocated during previous iteration.

Reply:
fixed